### DISCUSSION TITLE
Reduce Generics - Hide implementation details.

### DISCUSSION THREAD

**Discussion by semtexzv** [libp2p/rust-libp2p#1976] [Ideas]
Hello, I've been lurking around libp2p for some time, doing some hacking, and have summarized my issues here.
If it interests anyone:

While the capabilities of current design are quite impressive, It seems to be overly complex for what it does. It should be possible to work at the application layer without having to interact with complex traits like `Transport` and `StreamMuxer`, and their associated types.

This makes it hard to modify the library (I've tried several times, and and just got discouraged by the insanity of having to specify 8 generic parameters in some places ).

Several ideas:

**1. Remove Substream type from `StreamMuxer`** 
  -  force every implementation to do its own bookkeeping for substreams, same way as the wrapper muxer does.
  - Or make each substream carry a reference to underlying stream as an `Arc<Mutex>>`
  - Generally make substreams an implementation detail. Protocols should only ever know about AsyncRead/Write

**2. Remove `ConnectionInfo` as a trait, replace it with `PeerId`**
  -  It seems that this is the only implementor of this trait
  -  In places, where it might be useful to have more data about peers, Introduce this in local way, don't pollute the network, Swarm, and all other traits if its not needed. 

**3. Use boxing in reasonable places.** 
  -  This is a general trend in rust, that nothing is ever boxed, everything resolved statically. While it might be slightly more performant in some cases, the use of complex traits and generics everywhere introduces tight coupling which means modifications are very hard to implement.
  - We already do a bunch of copying for each network message. One indirect call through trait object is nothing in comparison.

**5. Move away from custom built combinators for `Transport` and `Upgrade`** 
  - Try to design a simpler approach based on async-await futures.

**6. Make behaviors independent.** 
 -  Current approach of building a behavior out of arbitrary ones works on 1 level, (You can't reasonably compose IPFS behavior with others in your behaviour, since then you can't use `Kademlia` behaviour yourself). Structuring this as a hierarchy is not correct. This is a set.
 - Introduce a separate control flow construct for each behaviour, and run them in parallel. Maybe a design with future for each behavior, where the future will:
      a. Listen for network messages, 
      b. Listen for Internal messages
      c. Emit network messages
      d. Emit internal mesages to other behaviours

My goal is not to just be negative, but to provide feedback from working with the library, and some directions, in which I'd like to move. I'm going to try to move forward on some of these myself, but I'm not sure how far I'll be able to go.

Let me know whether you're open to some of these changes, I'd be glad to do some work in this direction, I think it'd enable more features (like Relay)

**Comment by mxinden** [comment]
> My goal is not to just be negative, but to provide feedback from working with the library, and some directions, in which I'd like to move.

I very much appreciate your constructive actionable feedback. That said I don't think I am the right person to respond to most of this feedback.

### 1. Remove Substream type from StreamMuxer

Not the right person to answer this one.

### 2. Remove ConnectionInfo as a trait, replace it with PeerId

Off the top of my head I agree, but again, I don't think I am the right person here.

### 3. Use boxing in reasonable places.

I agree that dynamic dispatching should not be avoided at all costs and I also think that this mindset is reflected in the recent rust-libp2p changes, e.g. see https://github.com/libp2p/rust-libp2p/pull/1794. In case you have specific components in mind that would benefit from dynamic dispatching, suggestions are most welcome.

### 5. Move away from custom built combinators for Transport and Upgrade

Can you expand on how the current design is restricting you?

### 6. Make behaviors independent.

Yes, the fact that behaviours have to be constructed as hierarchies does make composition hard, your _IPFS Kademlia example_ being one such case. One could run multiple Kademlia instances within the `NetworkBehaviour` hierarchy, but given that Kademlia is rather resource heavy this is not a great trait-off. The same approach of duplication would work out for light-weight behaviours such as [libp2p-request-response](https://github.com/libp2p/rust-libp2p/tree/master/protocols/request-response).

Instead of designing deeply nested behaviour hierarchies, one can already have a flat hierarchy and a single root behaviour passing messages between the sub behaviours. This is a bit similar to your "separate control flow construct" suggestion, just in a synchronous fashion.

> Introduce a separate control flow construct for each behaviour, and run them in parallel.

The problem I see with structuring behaviours as independent components running in parallel is the difficulty of concurrency. Reasoning about a synchronous behaviour hierarchy with a single control flow is, in my eyes, easier compared to a set of behaviours running concurrently or even in parallel.

---

> I think it'd enable more features (like Relay)

Can you expand on how your suggestions would simplify the implementation of libp2p-relay (https://github.com/libp2p/rust-libp2p/pull/1134/)?

**Comment by tomaka** [comment]
> 2. Remove ConnectionInfo as a trait, replace it with PeerId

This genericity came from the fact that, in Substrate/Polkadot, we wanted to pass to the upper layers the ephemeral public key used during the encryption handshake, as a way to add a per-connection key.
The idea was abandoned, and we could indeed remove this.

> 3. Use boxing in reasonable places.

To me the problem of boxing isn't performances, but about having to decide whether to use `Box<dyn Trait>`, or `Box<dyn Trait + Send>`, or `Box<dyn Trait + Send + 'static>`, or `Box<dyn Trait + Send + 'a>` and have an `'a` on the owner, and so on.
If you use a generic parameter instead, this isn't a problem.

While this wasn't necessarily obvious in the past, it is now clearer that asynchronous I/O in Rust almost always uses ` + Send + 'static`. I agree that we could therefore use more boxing.

> 5. Move away from custom built combinators for Transport and Upgrade

The issue in doing this is, as often, the lack of existential types in the Rust language.

If you use non-temporary closures or futures, then you most likely have to store these closures or futures in a struct or enum at some point. There are two ways to do that:

- Adding a generic parameter to that struct/enum. But since the type of a closure/future isn't expressible, you can't actually write down the type of your struct/enum anymore. [Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f54bda758b1ff93c71949f79966ea86). This cascades into having to put generics everywhere, and causes even more frustration.
- Box the closure/future. But then you have the same problem as before, in that you have to decide whether the closure/future is forced to be `Send` and `'static`.

**Comment by thomaseizinger** [comment]
> Box the closure/future. But then you have the same problem as before, in that you have to decide whether the closure/future is forced to be `Send` and `'static`.

Isn't most stuff already required to be `Send + 'static` anyway?

- `NetworkBehaviour` is: https://github.com/libp2p/rust-libp2p/blob/30de0b4d644dac6dbe0cf6ac8c6498ebaf04b40d/swarm/src/behaviour.rs#L55
- `ProtocolsHandler` is: https://github.com/libp2p/rust-libp2p/blob/9365be711fd8b7e38a039c55d2a8650d76307411/swarm/src/protocols_handler.rs#L102
- all other type parameters of the `Swarm` are: https://github.com/libp2p/rust-libp2p/blob/dc56d44edb09e64cd0abaa10043897bdb4c91a44/swarm/src/lib.rs#L331-L335
- `Transport` itself isn't but the only way to construct a `Swarm` is with a boxed transport which requires `Send + 'static`: https://github.com/libp2p/rust-libp2p/blob/dc56d44edb09e64cd0abaa10043897bdb4c91a44/swarm/src/lib.rs#L338-L344
https://github.com/libp2p/rust-libp2p/blob/dc56d44edb09e64cd0abaa10043897bdb4c91a44/core/src/transport/boxed.rs#L27-L29

Doesn't this imply that the decision has already kind of be made? :D

**Comment by thomaseizinger** [comment]
> This genericity came from the fact that, in Substrate/Polkadot, we wanted to pass to the upper layers the ephemeral public key used during the encryption handshake, as a way to add a per-connection key.
> The idea was abandoned, and we could indeed remove this.

I gave this a spin here: https://github.com/libp2p/rust-libp2p/pull/1813
